File: packages/apl/python/src/apl_parser.py
# apl_parser.py
"""
Universal Prompt Language (APL) Parser
 
This module provides a parser for APL (A Programming Language) workflow files.
It reads a text file containing workflow steps and extracts the pre, prompt, and post phases.

Usage:
```python
from apl_parser import parse_apl
ast = parse_apl(open("workflow.apl").read())
print(ast["greet"].prompt.user)
```
"""

import re
from dataclasses import dataclass, field
from typing import Dict, List, Tuple, Final

# ── pre-compiled patterns ─────────────────────────────────────────────────────
PHASE_RE: Final  = re.compile(r'^\s*#\s*(pre|prompt|post)\s*:\s*(.*?)\s*$', re.I)
SUB_RE:   Final  = re.compile(r'^\s*##\s*(system|user)\s*(?:[:\-\s]*)$', re.I)
DASH_RE:  Final  = re.compile(r'^\s*---\s*$')
NUM_RE:   Final  = re.compile(r'^-?\d+(?:\.\d+)?$')

@dataclass
class Prompt:
    system: str = ''
    user:   str = ''

@dataclass
class Step:
    name: str
    slug: str
    pre_meta: Dict[str, object] = field(default_factory=dict)
    pre_code: str = ''
    prompt: Prompt = field(default_factory=Prompt)
    post_code: str = ''

# ── helpers ───────────────────────────────────────────────────────────────────
def _slug(s: str) -> str:
    return re.sub(r'[^a-z0-9]+', '_', s.lower()).strip('_')

def _cast(val: str):
    val = val.strip()
    if val.startswith('[') and val.endswith(']'):
        return [x.strip() for x in val[1:-1].split(',') if x.strip()]
    if NUM_RE.match(val):
        return float(val) if '.' in val else int(val)
    return val

def _parse_meta(lines: List[str], i: int) -> Tuple[Dict[str, object], int]:
    meta: Dict[str, object] = {}
    i += 1                                            # skip leading ---
    while i < len(lines) and not DASH_RE.match(lines[i]):
        if ':' in lines[i]:
            k, v = lines[i].split(':', 1)
            meta[k.strip()] = _cast(v)
        i += 1
    return meta, i + 1                                # skip closing ---

# ── main ──────────────────────────────────────────────────────────────────────
def parse_apl(text: str) -> Dict[str, Step]:
    lines = text.splitlines()
    steps: Dict[str, Step] = {}
    i, n = 0, len(lines)

    while i < n:
        m = PHASE_RE.match(lines[i])
        if not m:                                     # unrelated line
            i += 1
            continue

        phase, raw_name = m.group(1).lower(), m.group(2).strip()
        step = steps.setdefault(raw_name, Step(raw_name, _slug(raw_name)))
        i += 1

        # ── PRE ────────────────────────────────────────────────────────────
        if phase == 'pre':
            if i < n and DASH_RE.match(lines[i]):            # meta block
                meta, i = _parse_meta(lines, i)
                step.pre_meta.update(meta)
            buf = []
            while i < n and not PHASE_RE.match(lines[i]):
                buf.append(lines[i]); i += 1
            step.pre_code = '\n'.join(buf).rstrip()

        # ── PROMPT ─────────────────────────────────────────────────────────
        elif phase == 'prompt':
            cur: str | None = None
            buf: List[str] = []
            def flush():
                if cur:
                    setattr(step.prompt, cur, '\n'.join(buf).rstrip())
            while i < n and not PHASE_RE.match(lines[i]):
                s = SUB_RE.match(lines[i])
                if s:
                    flush()
                    cur, buf = s.group(1).lower(), []
                    i += 1
                    continue
                buf.append(lines[i]); i += 1
            flush()

        # ── POST ───────────────────────────────────────────────────────────
        else:  # post
            buf = []
            while i < n and not PHASE_RE.match(lines[i]):
                buf.append(lines[i]); i += 1
            step.post_code = '\n'.join(buf).rstrip()

    return steps

File: packages/apl/typescript/src/apl_parser.ts
// apl_parser.ts
/**
 * Universal Prompt Language (APL) Parser
 *
 * This parser reads a simple text format for defining workflow steps
 * with pre- and post-processing code, and user/system prompts.
 *
 * Usage:
 * ```typescript
 * import { parseApl } from "./aplParser.ts";
 * const ast = parseApl(await Deno.readTextFile("workflow.apl"));
 * console.log(ast["greet"].prompt.system);
 * ```
 */

export interface Prompt {
  system: string;
  user: string;
}
export interface Step {
  name: string;
  slug: string;
  preMeta: Record<string, unknown>;
  preCode: string;
  prompt: Prompt;
  postCode: string;
}

const PHASE_RE = /^\s*#\s*(pre|prompt|post)\s*:\s*(.*?)\s*$/i;
const SUB_RE = /^\s*##\s*(system|user)\s*(?:[:\-\s]*)$/i;
const DASH_RE = /^\s*---\s*$/;
const NUM_RE = /^-?\d+(?:\.\d+)?$/;

const slug = (s: string) =>
  s
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "_")
    .replace(/^_+|_+$/g, "");

function cast(val: string): unknown {
  val = val.trim();
  if (val.startsWith("[") && val.endsWith("]"))
    return val
      .slice(1, -1)
      .split(",")
      .map((s) => s.trim())
      .filter(Boolean);
  if (NUM_RE.test(val))
    return val.includes(".")
      ? Number.parseFloat(val)
      : Number.parseInt(val, 10);
  return val;
}

function parseMeta(
  lines: string[],
  idx: number,
): [Record<string, unknown>, number] {
  const meta: Record<string, unknown> = {};
  idx++; // skip opening ---
  while (idx < lines.length && !DASH_RE.test(lines[idx])) {
    const pos = lines[idx].indexOf(":");
    if (pos > -1)
      meta[lines[idx].slice(0, pos).trim()] = cast(lines[idx].slice(pos + 1));
    idx++;
  }
  return [meta, idx + 1]; // skip closing ---
}

export function parseApl(text: string): Record<string, Step> {
  const lines = text.split(/\r?\n/);
  const steps: Record<string, Step> = {};
  let i = 0;

  while (i < lines.length) {
    const m = lines[i].match(PHASE_RE);
    if (!m) {
      i++;
      continue;
    }

    const phase = m[1].toLowerCase();
    const name = m[2].trim();
    const step =
      steps[name] ??
      // biome-ignore lint/suspicious/noAssignInExpressions: <explanation>
      (steps[name] = {
        name,
        slug: slug(name),
        preMeta: {},
        preCode: "",
        prompt: { system: "", user: "" },
        postCode: "",
      });
    i++;

    if (phase === "pre") {
      if (i < lines.length && DASH_RE.test(lines[i])) {
        const [meta, next] = parseMeta(lines, i);
        Object.assign(step.preMeta, meta);
        i = next;
      }
      const buf: string[] = [];
      while (i < lines.length && !PHASE_RE.test(lines[i])) buf.push(lines[i++]);
      step.preCode = buf.join("\n").trimEnd();
    } else if (phase === "prompt") {
      let cur: "system" | "user" | null = null;
      let buf: string[] = [];
      const flush = () => {
        if (cur) step.prompt[cur] = buf.join("\n").trimEnd();
      };
      while (i < lines.length && !PHASE_RE.test(lines[i])) {
        const s = lines[i].match(SUB_RE);
        if (s) {
          flush();
          cur = s[1].toLowerCase() as "system" | "user";
          buf = [];
          i++;
          continue;
        }
        buf.push(lines[i++]);
      }
      flush();
    } else {
      // post
      const buf: string[] = [];
      while (i < lines.length && !PHASE_RE.test(lines[i])) buf.push(lines[i++]);
      step.postCode = buf.join("\n").trimEnd();
    }
  }
  return steps;
}
